cpp/oneapi/dal/algo/decision_forest/backend/gpu/feature_type_helper_dpc.cpp

```diff
@@ -0,0 +1,329 @@
+    requiredMem +=
+        sizeof(Bin) * row_count * column_count; // data vs ftrs bin map table (full_data_nd)
+    requiredMem +=
+        sizeof(Float) * column_count * std::min(max_bins, row_count); // bin_borders for each column
+
+    return requiredMem;
+}
+
+template <typename Float, typename Bin, typename Index>
+sycl::event indexed_features<Float, Bin, Index>::operator()(
+    const table& tbl,
+    const dal::backend::event_vector& deps) {
+    sycl::event::wait_and_throw(deps);
+
+    const auto data_nd_ =
+        pr::flatten_table<Float, row_accessor>(queue_, tbl, sycl::usm::alloc::device);
+
+    if (tbl.get_row_count() > de::limits<Index>::max()) {
+        throw domain_error(dal::detail::error_messages::invalid_range_of_rows());
+    }
+    if (tbl.get_column_count() > de::limits<Index>::max()) {
+        throw domain_error(dal::detail::error_messages::invalid_range_of_columns());
+    }
+
+    row_count_ = de::integral_cast<Index>(tbl.get_row_count());
+    column_count_ = de::integral_cast<Index>(tbl.get_column_count());
+    total_bins_ = 0;
+
+    //allocating buffers
+    full_data_nd_ =
+        pr::ndarray<Bin, 2>::empty(queue_, { row_count_, column_count_ }, sycl::usm::alloc::device);
+    bin_offsets_nd_ =
+        pr::ndarray<Index, 1>::empty(queue_, { column_count_ + 1 }, sycl::usm::alloc::device);
+
+    entries_.resize(column_count_);
+
+    auto values_nd = pr::ndarray<Float, 1>::empty(queue_, { row_count_ }, sycl::usm::alloc::device);
+    auto indices_nd =
+        pr::ndarray<Index, 1>::empty(queue_, { row_count_ }, sycl::usm::alloc::device);
+    std::vector<pr::ndarray<Bin, 1>> column_bin_vec_;
+    column_bin_vec_.resize(column_count_);
+
+    for (Index i = 0; i < column_count_; i++) {
+        column_bin_vec_[i] =
+            pr::ndarray<Bin, 1>::empty(queue_, { row_count_ }, sycl::usm::alloc::device);
+    }
+
+    pr::radix_sort_indices_inplace<Float, Index> sort{ queue_ };
+
+    sycl::event cross_event;
+
+    for (Index i = 0; i < column_count_; i++) {
+        auto extract_event = extract_column(data_nd_, values_nd, indices_nd, i, { cross_event });
+        auto sort_event = sort(values_nd, indices_nd, { extract_event });
+        auto cross_event =
+            compute_bins(values_nd, indices_nd, column_bin_vec_[i], entries_[i], { sort_event });
+    }
+
+    cross_event.wait_and_throw();
+
+    auto bin_offsets_nd_host = bin_offsets_nd_.to_host(queue_);
+    auto bin_offsets = bin_offsets_nd_host.get_mutable_data();
+
+    Index total = 0;
+    for (Index i = 0; i < column_count_; i++) {
+        cross_event =
+            store_column(column_bin_vec_[i], full_data_nd_, i, column_count_, { cross_event });
+        bin_offsets[i] = total;
+        entries_[i].offset_ = total;
+        total += entries_[i].bin_count_;
+    }
+
+    bin_offsets[column_count_] = total;
+    total_bins_ = total;
+
+    bin_offsets_nd_ = bin_offsets_nd_host.to_device(queue_);
```

Comment:
Suggest to move to separate functions. Read more: http://principles-wiki.net/principles:single_level_of_abstraction